File: neetcode/problems/leetcode912/solution.py
import math
from typing import List


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        return self.mergeSort(nums, 0, len(nums) - 1)

    def mergeSort(self, nums: List[int], start: int, end: int):
        if start == end:
            return nums
        mid = math.floor((start + end) / 2)

        self.mergeSort(nums, start, mid)
        self.mergeSort(nums, mid + 1, end)

        return self.merge(nums, start, mid, end)

    def merge(self, nums: List[int], start: int, mid: int, end: int):
        left = nums[start : mid + 1]
        right = nums[mid + 1 : end + 1]

        left_index = 0
        right_index = 0
        index = start

        while left_index < len(left) and right_index < len(right):
            if left[left_index] <= right[right_index]:
                nums[index] = left[left_index]
                left_index += 1
            else:
                nums[index] = right[right_index]
                right_index += 1
            index += 1

        while left_index < len(left):
            nums[index] = left[left_index]
            left_index += 1
            index += 1

        while right_index < len(right):
            nums[index] = right[right_index]
            right_index += 1
            index += 1

        return nums

File: neetcode/problems/leetcode912/solution.js
const sortArray = nums => {
	return mergeSort(nums, 0, nums.length - 1);
};

const mergeSort = (nums, start, end) => {
	if (start === end) return nums;

	const mid = Math.floor((start + end) / 2);

	mergeSort(nums, start, mid);
	mergeSort(nums, mid + 1, end);

	return merge(nums, start, mid, end);
};

const merge = (nums, start, mid, end) => {
	const left = nums.slice(start, mid + 1);
	const right = nums.slice(mid + 1, end + 1);

	let li = 0;
	let ri = 0;
	let i = start;

	while (li < left.length && ri < right.length) {
		if (left[li] <= right[ri]) {
			nums[i] = left[li];
			li++;
		} else {
			nums[i] = right[ri];
			ri++;
		}

		i++;
	}

	while (li < left.length) {
		nums[i] = left[li];
		li++;
		i++;
	}

	while (ri < right.length) {
		nums[i] = right[ri];
		ri++;
		i++;
	}

	return nums;
};

function findUniq(arr) {
	const lettersScores = {};

	arr.forEach((str, i) => {
		for (const letter of str.toLowerCase()) {
			if (lettersScores[letter]) lettersScores[letter][0]++;
			else lettersScores[letter] = [1, i];
		}
	});

	let lowestScore = Infinity;
	let idx = -1;

	for (const letter in lettersScores) {
		const [count, index] = lettersScores[letter];

		if (count < lowestScore) {
			lowestScore = count;
			idx = index;
		}
	}

	return arr[idx];
}

console.log(findUniq(['foobar', 'barfo', 'fobara', '   ', 'fobra', 'oooofrab']));
